### ISSUE TITLE
figure out DinD + podutils

### ISSUE THREAD

**Issue by BenTheElder** [kubernetes/test-infra#8780]
This is a specific issue to track the need for Docker in Docker with the podutils. I am also filing a related but more general issue for the podutils and init / forking that may wind up solving this one. xref: https://github.com/kubernetes/test-infra/issues/8779

In short, with the old bootstrap + scenario scripts and their images we have support for using docker at runtime in prowjobs, with the podutils we do not. We need to sort this out. A complication here is that the podutils now control the entrypoint, we can run a similar script under the entrypoint but before the actual test command, but that leaves other questions, like how we will handle zombie processes...

xref: https://github.com/kubernetes/test-infra/pull/8763#issuecomment-407097014, https://github.com/kubernetes/test-infra/issues/7844#issuecomment-407069769

**Comment by BenTheElder** [comment]
@stevekuznetsov I don't think that will work properly due to volume mounts. See https://github.com/kubernetes/test-infra/issues/8779#issuecomment-407121384

Seperately from DinD, most things we run fork, avoiding zombies is something to handle more generally. What needs to happen here specifically is a different snippet to boot docker (instead of having it in the images/bootstrap entrypoint.sh)

**Comment by BenTheElder** [comment]
We should be able to refactor out just the DinD parts from `images/bootstrap/runner` and let the pdoutils job be `entrypoint test_cmd` where `test_cmd` is something like `dind_runner actual-test-command`. 

This just reminded me that zombies are also coming to eat our nodes. For that something like `/bin/sh entrypoint test_cmd` would probably be sufficient.

I'll handle the refactor bit now and circle back on zombies. Kubernetes may also soon have better ability to manage this problem as well... (it does have some alpha gated enhancements now IIRC) .. we're not the only ones with "the PID1 problem".

**Comment by stevekuznetsov** [comment]
> I'll handle the refactor bit now and circle back on zombies. Kubernetes may also soon have better ability to manage this problem as well

I'd check where the `pause` container work is, that was supposed to handle the issue IIRC

**Comment by BenTheElder** [comment]
It looks like we are waiting on https://github.com/kubernetes/kubernetes/issues/1615 before the pause container can reap zombies. It has some logic already, but currently it doesn't work as the pod doesn't share a PID namespace yet. 
The last update is that 1.10 had this as an alpha feature https://github.com/kubernetes/kubernetes/issues/1615#issuecomment-386976958, until just two hours ago A PR opened to push it to beta https://github.com/kubernetes/kubernetes/pull/66507 🤞 

It will be a while before Prow is running on this though, as even 1.11 has it as alpha, 1.12 _may_ have it as beta.

**Comment by stevekuznetsov** [comment]
> It will be a while before Prow is running on this though

You are stock GKE, right? I think we may be running with it right now on ours, but we fiddle with cluster flags a lot.

**Comment by stevekuznetsov** [comment]
> let the pdoutils job be entrypoint test_cmd where test_cmd is something like dind_runner actual-test-command.

Are there benefits from integrating more tightly for classes of jobs (DinD, jobs that want VMs, stuff like that) or do we just want to black-box everything and keep it as run by `entrypoint`?

**Comment by BenTheElder** [comment]
Stock GKE does not enable alpha features to my knowledge. We could almost certainly sidestep this but it would be hacky.

> Are there benefits from integrating more tightly for classes of jobs (DinD, jobs that want VMs, stuff like that) or do we just want to black-box everything and keep it as run by entrypoint?

We should keep it flexible. I don't think we want to support VMs etc as part of PRow. I do think we need to support forking, as tons of useful utilities fork. This issue is really for tracking that our images also need an update because they assume they can do $magic in the entrypoint before running bootstrap.

**Comment by verb** [comment]
FYI, GKE supports [alpha clusters](https://cloud.google.com/kubernetes-engine/docs/concepts/alpha-clusters), but they're only suitable for testing.

@stevekuznetsov Are you running with PodShareProcessNamespace=true? I haven't received much feedback on this feature. I'd like to gather a few examples of people using it successfully.

**Comment by BenTheElder** [comment]
Yeah, alpha clusters are not a suitable option for Prow. Some end to end testing uses them though.

**Comment by BenTheElder** [comment]
In theory after #8874 a job should be able to use sysv DinD + podutils relatively fine. I'll experiment with this or help someone port a job to it soonish. perhaps @MHBauer

**Comment by spiffxp** [comment]
/milestone v1.13
I actually have no clue here, I'm going based off of the priority label, otherwise I would have punted from the milestone.  Any updates?

**Comment by BenTheElder** [comment]
pod-utils + dind works with our sysv dind, it's not quite ideal though, but we should probably track options for not hijacking ENTRYPOINT in another issue.
